File: java/src/main/java/deltix/luminary/InterfaceDef.java
package deltix.luminary;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public class InterfaceDef extends ClassOrInterfaceDef<InterfacePropertyDef, TypeInterface> {
    private final TypeInterface type;
    private final List<InterfaceMethodDef> methods = new ArrayList<>();

    public InterfaceDef(@NotNull TypeScope parent, @NotNull String name, @Nullable List<String> comments) {
        super(parent, name, comments);
        this.type = new TypeInterface(this);

        parent.getDefinedInterfaces().add(this);
    }

    public InterfaceDef(@NotNull TypeScope parent, @NotNull String name) {
        this(parent, name, null);
    }

    @NotNull
    @Override
    public TypeInterface getType() {
        return type;
    }

    @NotNull
    public List<InterfaceMethodDef> getMethods() {
        return methods;
    }
}

File: csharp/main/InterfaceDef.cs
using System;
using System.Collections.Generic;
using JetBrains.Annotations;

namespace Deltix.Luminary
{
	/// <summary>
	/// Definition of the Luminary interface.
	/// </summary>
	public class InterfaceDef : ClassOrInterfaceDef<InterfacePropertyDef, TypeInterface>
	{
		public InterfaceDef([NotNull] ITypeScope parent, [NotNull] String name, [CanBeNull, ItemNotNull] List<String> comments = null)
			: base(name, parent, comments)
		{
			Type = new TypeInterface(this);

			FullName = parent is FileDef ? Name : $"{((ItemDef) parent).FullName}.{Name}";
			Methods = new List<InterfaceMethodDef>();

			parent.DefinedInterfaces.Add(this);
		}

		/// <summary>
		/// List of methods defined by this interface.
		/// </summary>
		[NotNull]
		public List<InterfaceMethodDef> Methods { get; }

		public override TypeInterface Type { get; }

		/// <summary>
		/// Full name of the interface.
		/// </summary>
		public override String FullName { get; }
	}
}
